### PR TITLE
Explore: use numeric ids for panel ids when building query transactions

### LINKED ISSUE #64515: Explore: Use number for panelIds when building query transaction
a `DataQueryRequest` has an optional field, `panelId` defined as a number: https://github.com/grafana/grafana/blob/46ab9e9252e48eba7f6c5acbd76cf94b55db9b6b/packages/grafana-data/src/types/datasource.ts#L510

when a request is created in explore-mode, it also gets a panelId value, but it is a string:
https://github.com/grafana/grafana/blob/46ab9e9252e48eba7f6c5acbd76cf94b55db9b6b/public/app/core/utils/explore.ts#L146-L148 (it's an UUID-ish string value).

this causes problems for consumers of `.panelId`, who assume it is a number.

(please note that this value is also sent in every backend-datasource-ajax-request, so the number of consumers is not limited to the javascript-in-browser)

### LINKED ISSUE #70427: Explore: detect id collision when generating panes IDs
Probably very low probability, but should we detect collision with existing id?

_Originally posted by @ifrost in https://github.com/grafana/grafana/pull/69692#discussion_r1235138425_
            
in https://github.com/grafana/grafana/pull/69692 we introduced a new way of generating Explore Ids. Although the probability of it happening should be quite low, we should check whether - upon creating a new pane - the generated id matches and already opened pane, and if so generate a new one.

### PR DESCRIPTION
**Pull request by Elfo404**
for context: https://raintank-corp.slack.com/archives/CHKLEJ668/p1702997323296379
makes use of a custom nanoid alphabet (`01234567890abcdefghijklmnopqrstuwxyz`) to generate explore ids, so that any explore id can be parsed as a unique integer when generating panelIds when building transactions.
When navigating to explore from an external URL, if pane ids in the URL are not part of the alphabet, are not 3 chars long or are only numeric, new valid ids are generated.

additionally, prevents Ids duplication when opening split panes

Fixes #70427
Fixes #64515

### CODE REVIEW

**Review by gelicia — APPROVED** [review]

**Comment by Elfo404** [comment]
not really, there shouldn't be any behavior change, the reason i made the ids composed only of lowercase letters is to use `parseInt(id, 36)` to generate a unique numeric id based on the pane id. by using only lowercase letters and number the alphabet used to generate IDs is composed only of 36 characters, which in turn ensure that for every possible pane id there will be a unique numeric id used in `panelId` when issuing queries. the reason why i had to remove uppercase letters is because the `radix` parameter in `parseInt` only accepts integers between `2` and `36`.

**Comment by ifrost** [comment]
Should it also say "Fixes #64515"?

File: public/app/core/utils/explore.ts [hunk 1]
```diff
@@ -52,7 +55,13 @@ export interface GetExploreUrlArguments {
 }
 
 export function generateExploreId() {
-  return nanoid(3);
+  while (true) {
+    const id = nanoid(3);
+
+    if (!/^\d+$/.test(id)) {
```
Comment (ifrost) [inline public/app/core/utils/explore.ts:61]: Why do we need it? 🤔

Comment (Elfo404) [reply public/app/core/utils/explore.ts:61]: yes, numeric-only panel ids will lose the ordering when parsed, this prevents that

**Pull request merged**